File: java2cppc/src/matsushima/common/cc/DefinitionReader.java
package matsushima.common.cc;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.logging.Log;

/**
 * 解析定義読み込み。
 * 
 * @author  0.0.1.001 2010/05/20 matsushima
 * @version 0.0.1.001 2010/05/20 matsushima
 * @since   0.0.1.001 2010/05/20 matsushima
 * 
 * $Id: DefinitionReader.java,v 1.2 2010/05/23 18:36:43 matsushima Exp $
 * $Log: DefinitionReader.java,v $
 * Revision 1.2  2010/05/23 18:36:43  matsushima
 * 検索
 *
 */
public abstract class DefinitionReader {

    /**
     * 行の key と value を解析する。
     * 
     * @param line
     * @param index
     * @param comment
     * @param p_key
     * @param p_value
     * @return
     */
    protected abstract boolean parseKeyAndValueLine(
            String line,
            int index,
            String comment,
            Pattern p_key,
            Pattern p_value);

    /**
     * 定義を読み込む。
     * 
     * @param in
     * @throws IOException
     */
    public void readDefinition(
            final Reader in,
            final Log log_readdef)
            throws IOException {
        BufferedReader bin = null;
        try {
            // read
            bin = new BufferedReader(in);
            // read key, value, comment pattern
            Pattern p_key = null;
            Pattern p_value = null;
            Pattern p_line_comment = null;
            Pattern p_block_comment_in = null;
            Pattern p_block_comment_out = null;
            String line = bin.readLine();
            Pattern p = Pattern.compile(line);
            while (null != (line = bin.readLine())) {
                Matcher m = p.matcher(line); // [key : ]value value ... // line_comment
                boolean b = m.lookingAt();
                if (b) {
                    String key = m.group(1);
                    String value = m.group(2);
log_readdef.debug(key + ": " + value);
                    if ("key".equals(key)) {
                        p_key = Pattern.compile(value);
                    } else if ("value".equals(key)) {
                        p_value = Pattern.compile(value);
                    } else if ("LineComment".equals(key)) {
                        p_line_comment = Pattern.compile(value);
                    } else if ("BlockCommentIn".equals(key)) {
                        p_block_comment_in = Pattern.compile(value);
                    } else if ("BlockCommentOut".equals(key)) {
                        p_block_comment_out = Pattern.compile(value);
                    }
                    if ((null != p_key) && (null != p_value) && (null != p_line_comment)
                            && (null != p_block_comment_in) && (null != p_block_comment_out)) {
                        break;
                    }
                }
            }
            // read key, value, comment
            boolean in_block_comment = false;
            while (null != (line = bin.readLine())) {
                int index = 0;
log_readdef.debug("< line: " + line);
                // block_comment
                if (! in_block_comment) {
                    Matcher m = p_block_comment_in.matcher(line); // [...][/* block_comment]
                    boolean b = m.lookingAt();
                    if (b) {
                        in_block_comment = true;
                        line = m.group(1);
                        String comment = m.group(2);
log_readdef.debug("new line: " + line);
log_readdef.debug("comment: " + comment);
                    }
                } else {
                    Matcher m = p_block_comment_out.matcher(line); // [block_comment */][...]
                    boolean b = m.lookingAt();
                    if (b) {
                        in_block_comment = false;
                        line = m.group(2);
                        String comment = m.group(1);
log_readdef.debug("new line: " + line);
log_readdef.debug("comment: " + comment);
                    } else {
log_readdef.debug("in_block_comment");
                        continue;
                    }
                }
                // line_comment
                String comment = null;
                Matcher m = p_line_comment.matcher(line); // key : value value ... [// line_comment]
                boolean b = m.lookingAt();
                if (b) {
                    line = m.group(1);
                    comment = m.group(2);
log_readdef.debug("new line: " + line);
log_readdef.debug("comment: " + comment);
                }
                // key, value
                if (! parseKeyAndValueLine(line, index, comment, p_key, p_value)) {
                    break;
                }
            }
        } finally {
            try {
                if (null != bin) {
                    bin.close();
                }
            } finally {
                in.close();
            }
        }
    }
}

File: srcconv/DefinitionReader.cpp
#include "StdAfx.h"
#include "DefinitionReader.h"

DefinitionReader::DefinitionReader(void)
{
}


DefinitionReader::~DefinitionReader(void)
{
}

/**
 * 定義を読み込む。
 * 
 * @param in
 * @throws IOException
 */
void DefinitionReader::readDefinition(
	Reader& in,
	Log& log_readdef)
{
	// read key:value, key, value, xxxcommentxxx pattern
	std::tr1::regex pattern_key;
	std::tr1::regex pattern_value;
	std::tr1::regex pattern_line_comment;
	std::tr1::regex pattern_block_comment_in;
	std::tr1::regex pattern_block_comment_out;
	std::tr1::smatch match;
	std::string line;
	if (NULL == in.readLine(line))
	{
		return;
	}
	std::tr1::regex pattern_key_value(line);
	int pattern_flags = 0;
	while (NULL != in.readLine(line))
	{
		log_readdef.debug("< " + line);
		if (std::tr1::regex_search(line, match, pattern_key_value)) // [key : ]value value ... // line_comment
		{
			std::string key = match.str(1);
			std::string value = match.str(2);
			log_readdef.debug("> " + key + ": " + value);
			if (0 == key.compare("key"))
			{
				pattern_key = value;
				pattern_flags |= 1;
			}
			else if (0 == key.compare("value"))
			{
				pattern_value = value;
				pattern_flags |= 2;
			}
			else if (0 == key.compare("LineComment"))
			{
				pattern_line_comment = value;
				pattern_flags |= 4;
			}
			else if (0 == key.compare("BlockCommentIn"))
			{
				pattern_block_comment_in = value;
				pattern_flags |= 8;
			}
			else if (0 == key.compare("BlockCommentOut"))
			{
				pattern_block_comment_out = value;
				pattern_flags |= 16;
			}
			if (31 == pattern_flags)
			{
				break;
			}
		}
	}
	// read key, value, comment
	bool in_block_comment = false;
	std::string comment;
	while (NULL != in.readLine(line))
	{
		log_readdef.debug("< " + line);
		std::string::const_iterator line_begin = line.cbegin();
		std::string::const_iterator line_end = line.cend();
		// block_comment
		if (!in_block_comment)
		{
			if (std::tr1::regex_search(line_begin, line_end, match, pattern_block_comment_in, std::tr1::regex_constants::match_continuous)) // [...][/* block_comment]
			{
				in_block_comment = true;
				comment = match.str(2);
				line_end = line_begin + match.position(2);
				log_readdef.debug("new line(comment in): " + std::string(line_begin, line_end));
				log_readdef.debug("comment: " + comment);
			}
		}
		else
		{
			if (std::tr1::regex_search(line_begin, line_end, match, pattern_block_comment_out, std::tr1::regex_constants::match_continuous)) // [block_comment */][...]
			{
				in_block_comment = false;
				comment = match.str(1);
				line_begin += match.position(1) + match.length(1);
				log_readdef.debug("new line(comment out): " + std::string(line_begin, line_end));
				log_readdef.debug("comment: " + comment);
			}
			else
			{
				log_readdef.debug("in_block_comment");
				continue;
			}
		}
		// line_comment
		comment.clear(); // block comment は捨てる
		if (std::tr1::regex_search(line_begin, line_end, match, pattern_line_comment, std::tr1::regex_constants::match_continuous)) // key : value value ... [// line_comment]
		{
			comment = match.str(2);
			line_end = line_begin + match.position(2);
			log_readdef.debug("new line(line comment): " + std::string(line_begin, line_end));
			log_readdef.debug("comment: " + comment);
		}
		// key and/or value
		// 行の key と value を解析する。
		if (!parseKeyAndValueLine(std::string(line_begin, line_end), 0, comment, pattern_key, pattern_value))
		{
			break;
		}
	}
}
